### PR TITLE
Upgrade TypeScript

### PR DESCRIPTION
**Pull request by bobjflong**

### CODE REVIEW

File: source/index.ts [hunk 1]
```diff
@@ -226,6 +223,7 @@ export default class PQueue<QueueType extends Queue<RunFunction, EnqueueOptionsT
 	/**
 	Adds a sync or async task to the queue. Always returns a promise.
 	*/
+	// eslint-disable-next-line @typescript-eslint/prefer-readonly-parameter-types
 	async add<TaskResultType>(fn: Task<TaskResultType>, options: Partial<EnqueueOptionsType> = {}): Promise<TaskResultType> {
```
Comment (bobjflong) [inline source/index.ts]: I think it's tricky to use `Readonly` types here because the `{}` assignment.

File: package.json [hunk 2]
```diff
@@ -48,20 +48,20 @@
...
 		"delay": "^4.2.0",
-		"eslint-config-xo-typescript": "^0.16.0",
+		"eslint-config-xo-typescript": "^0.27.0",
 		"in-range": "^2.0.0",
 		"nyc": "^14.0.0",
 		"random-int": "^2.0.0",
 		"time-span": "^3.1.0",
 		"ts-node": "^8.3.0",
-		"typescript": "3.7.2",
+		"typescript": "^3.8.3",
```
Comment (sindresorhus) [inline package.json]: This should stay locked. TypeScript breaks stuff all the time in minor releases.

File: source/index.ts [hunk 3]
```diff
@@ -10,7 +10,7 @@ type Task<TaskResultType> =
 	| (() => PromiseLike<TaskResultType>)
 	| (() => TaskResultType);
 
-const empty = (): void => {};
+const empty = (): void => { /* do nothing */ };
```
Comment (sindresorhus) [inline source/index.ts]: Use a ESLint disable comment instead of this.

File: source/index.ts [hunk 4]
```diff
@@ -53,27 +53,25 @@ export default class PQueue<QueueType extends Queue<RunFunction, EnqueueOptionsT
...
 			intervalCap: Infinity,
 			interval: 0,
 			concurrency: Infinity,
 			autoStart: true,
 			queueClass: PriorityQueue,
-			...options
-		// TODO: Remove this `as`.
-		} as Options<QueueType, EnqueueOptionsType>;
+			...passedOptions
+		} as any;
```
Comment (sindresorhus) [inline source/index.ts]: Is this really any different/better than what was already there?

File: source/index.ts (hunk 1 above)
Comment (sindresorhus) [reply source/index.ts]: How so? Can you elaborate? If it's not possible to use readonly here, it sounds like a the linter rule should be adjusted to ignore this case.

File: source/index.ts (hunk 4 above)
Comment (bobjflong) [reply source/index.ts]: It was just to avoid disabling the lint rule. I'll revert this one too.

File: source/index.ts (hunk 1 above)
Comment (bobjflong) [reply source/index.ts]: If you type `options` as 

```
Readonly<Partial<EnqueueOptionsType
```

Then this type error is returned:

```
Type '{}' is not assignable to type 'Readonly<Partial<EnqueueOptionsType>>'.ts(2322)
```

And the linter _does_ want `options` to be Readonly afaict:

```
source/index.ts:229:58
  ✖  229:58  Parameter should be a read only type
```

Not an expert, but this indeed could be a linting issue I think. We shouldn't require options to be readonly in this case.

Comment (bobjflong) [reply source/index.ts]: Let me know what you'd prefer in this case. Happy to look into that linting thing too if it's confirmed an issue.

Comment (sindresorhus) [reply source/index.ts]: If TypeScript doesn't allow it, it's a linter bug. Can you open an issue here: https://github.com/typescript-eslint/typescript-eslint ?

**Comment by sindresorhus** [comment]
👌

**Comment by bobjflong** [comment]
@sindresorhus  Without the `.eslintignore` file, this fails to build for me locally:

```
[p-queue] npm run test

> p-queue@6.3.0 test /Users/bob/src/p-queue
> xo && npm run build && nyc ava


  bench.ts
  ✖  Parsing error: "parserOptions.project" has been set for @typescript-eslint/parser.
The file does not match your project config: bench.ts.
The file must be included in at least one of the projects provided.      

  test/test.ts
  ✖  Parsing error: "parserOptions.project" has been set for @typescript-eslint/parser.
The file does not match your project config: test/test.ts.
The file must be included in at least one of the projects provided.  

```

**Comment by sindresorhus** [comment]
I fixed that by upgrading the `xo` dependency.

**Comment by bobjflong** [comment]
Ah, cool

**Pull request merged**